README.md
[![Build Status](https://travis-ci.org/G-Node/nix.svg?branch=master)](https://travis-ci.org/G-Node/nix)
[![Build status](https://ci.appveyor.com/api/projects/status/1qlcasjg2fpqotig/branch/master?svg=true)](https://ci.appveyor.com/project/G-Node/nix/branch/master)
[![Coverage](https://codecov.io/gh/G-Node/nix/branch/master/graph/badge.svg)](https://codecov.io/gh/G-Node/nix)

---

![nix_logo](docs/nix_logo.png "NIX")

About NIX
---------

The *NIX* project started as an initiative within the
Electrophysiology Task Force a part of the
[INCF](http://www.incf.org/) Data sharing Program.  The *NIX* data
model allows to store fully annotated scientific dataset, i.e. the
data together with its metadata within the same container. Our aim is
to achieve standardization by providing a common/generic data
structure for a multitude of data types.  [See the wiki for more
information](https://github.com/G-Node/nix/wiki)


The current implementations store the actual data using the [HDF5](http://www.hdfgroup.org/) file format as a storage backend.

Documentation --- where to find which information?
--------------------------------------------------

Project documentation is split up into three parts:
1. Technical information: Please visit the GitHub repository [Wiki](https://github.com/G-Node/nix/wiki).
2. General Introduction and tutorial: Visit the project pages on [ReadTheDocs](https://nixio.readthedocs.io/en/latest/)
3. API documentation: Is part of the project pages on [ReadTheDocs](https://nixio.readthedocs.io/en/latest/api/library_root.html)

Further project related information:

- We have assembled a set of
 [tutorials](https://nixpy.readthedocs.io/en/latest/tutorial.html "Python Tutorial") and
 [demos](https://github.com/g-node/nix-demo "Jupiter notebooks demonstrating nix for various use-cases") using the python library.


The NIX ecosystem
-----------------

In this repository you find the C++ library for reading and writing NIX files.

**Language bindings**

We provide bindings in various languages:

- [Python bindings: *nixpy*](https://github.com/g-node/nixpy "Python library either as bindings or using h5py")
- [Matlab: *nix-mx*](https://github.com/g-node/nix-mx "Matlab language bindings, requires the C++ library")
- [Java: *nix-java*](https://github.com/g-node/nix-java "Java language bindings, requires the C++ library")

[IO class](https://github.com/python-neo-nixio) for the [NEO](http://neuralensemble.org/neo/) data model for spike train data.

**Tools**

- Viewer for *NIX* data files [nixView](https://github.com/bendalab/nixview "NixView - viewer for nix files")


Getting Started
---------------

Platform specific installation instructions can be found:

- [Linux](docs/install_linux.md)
- [macOS](docs/install_mac.md)
- [Windows](docs/install_win.md)


debian/changelog
nix (1.4.5-3) xenial; urgency=medium

  * version bump to 1.4.5

 -- Jan Grewe <jan.grewe@g-node.org>  Tue, 20 Apr 2021 13:52:00 +0200

nix (1.4.5-2) bionic; urgency=medium

  * version bump to 1.4.5

 -- Jan Grewe <jan.grewe@g-node.org>  Tue, 20 Apr 2021 14:18:00 +0200

nix (1.4.5-1) focal; urgency=medium

  * version bump to 1.4.5

 -- Jan Grewe <jan.grewe@g-node.org>  Tue, 20 Apr 2021 13:52:00 +0200

nix (1.4.4-4) focal; urgency=medium

  * version bump to 1.4.4

 -- Jan Grewe <jan.grewe@g-node.org>  Mon, 19 Apr 2021 14:31:00 +0200

nix (1.4.4-3) focal; urgency=medium

  * version bump to 1.4.4

 -- Jan Grewe <jan.grewe@g-node.org>  Fri, 16 Apr 2021 16:20:00 +0100

nix (1.4.4-2) bionic; urgency=medium

  * version bump to 1.4.4

 -- Jan Grewe <jan.grewe@g-node.org>  Fri, 16 Apr 2021 16:17:00 +0100

nix (1.4.4-1) xenial; urgency=medium
  * version bumb to 1.4.4
  * fixes installation of static build
  * allows to tag more flexibly: i.e. a Tag may now tag DataArrays that differ in dimensionality. Superfluous dimension information is ignored, missing information is added #801.
  * fixes a bug in the IndexOf methods of RangeDimensions #819
  * some internal modifications regarding error handling

 -- Jan Grewe <jan.grewe@g-node.org>  Thu,  Nov 2021 13:53:00 +0100

nix (1.4.3-3) bionic; urgency=medium

  * version bump to 1.4.3

 -- Jan Grewe <jan.grewe@g-node.org>  Thu, 01 Nov 2018 16:09:00 +0100

nix (1.4.3-2) xenial; urgency=medium

  * version bump to 1.4.3

 -- Jan Grewe <jan.grewe@g-node.org>  Thu, 01 Nov 2018 16:09:00 +0100

nix (1.4.3-1) trusty; urgency=medium

  * fix of compile issues with the latest hdf5 version
  * add data slicing method
  * improvements of retrieve data functionality: multiple slices an be retrieved at once
  * name and type filter functions accept regular expressions
  * add default arguments for appending dimensions
  * adds a file open flag that allows to force open a file skipping version and format checks
  * changes to static build
  * fixed a bug when 1D data was retrieved from a tag with 2D positions

 -- Jan Grewe <jan.grewe@g-node.org>  Thu, 01 Nov 2018 15:40:00 +0100

nix (1.4.1-2) xenial; urgency=medium

  * version bump to 1.4.1

 -- Jan Grewe <jan.grewe@g-node.org>  Fri, 20 Oct 2017 16:33:04 +0100

nix (1.4.1-1) trusty; urgency=medium

  * internal improvements (test coverage, fixed build warning, arm build working, improved string handling)
  * fixed and improved seraching of sections in metadata
  * add option to compress datasets
  * allow for non-si units at several places

 -- Jan Grewe <jan.grewe@g-node.org>  Fri, 20 Oct 2017 09:40:00 +0100

nix (1.4.0-2) xenial; urgency=medium

  * version bump to 1.4.0

 -- Jan Grewe <jan.grewe@g-node.org>  Tue, 06 Jun 2017 11:16:04 +0100

nix (1.4.0-1) trusty; urgency=medium

  * improved performance when adding entities to groups, or appending references and features
  * references and features can now be found also by name, not only by id
  * fixes a bug, that allowed appending of entities from foreign blocks
  * fixes a bug in the guess chunking algorithm when creating very small or empty datasets (led to chunk size of 1 and resulted in very low performance during read)
  * command line tool reports version numbers

 -- Jan Grewe <jan.grewe@g-node.org>  Tue, 06 Jun 2017 10:59:16 +0100

nix (1.3.2-2) xenial; urgency=medium

  * Version bump to 1.3.2

 -- Jan Grewe <jan.grewe@g-node.org>  Thu, 12 Jan 2017 15:40:16 +0100

nix (1.3.2-1) trusty; urgency=medium

 * Fix error in nix-tool that prevented dumping older format versions.
 * Add flush() method for persisting file at user-defined times.
 * Version set to 1.3.2

 -- Jan Grewe <jan.grewe@g-node.org> Thu, 12 Jan 2017 14:50:00 +0100

nix (1.3.1-1) xenial; urgency=medium

 * Set version to 1.3.1

 -- Jan Grewe <jan.grewe@g-node.org> Fri, 25 Nov 2016 08:40:00 +0100

nix (1.3.1-0) trusty; urgency=medium

 * Fix error of missing header
 * Improve bitfield2bool conversion
 * Version set to 1.3.1

 -- Jan Grewe <jan.grewe@g-node.org> Fri, 25 Nov 2016 08:33:00 +0100

nix (1.3.0-1) xenial; urgency=medium

 * Set to version 1.3.0

 -- Jan Grewe <jan.grewe@g-node.org> Wed, 16 Nov 2016 18:24:00 +0100

nix (1.3.0-0) trusty; urgency=medium

 * HDF5 file format increased to 1.1.0
 * Change way of handling booleans in hdf5
 * adds api-version information

 -- Jan Grewe <jan.grewe@g-node.org> Wed, 16 Nov 2016 18:21:00 +0100

nix (1.2.1-1)  xenial; urgency=medium

 * Set to version 1.2.1

 -- Jan Grewe <jan.grewe@g-node.org> Tue, 16 Nov 2016 18:02:00 +0100

nix (1.2.1-0) trusty; urgency=medium

 * Set to version 1.2.1
 * adds checks for file format version
 * adds utility functions for support of enum types in DataSet
 * updates documentation

 -- Jan Grewe <jan.grewe@g-node.org> Wed, 16 Nov 2016 17:58:22 +0100

nix (1.2.0-1)  xenial; urgency=medium

 * Set to version 1.2.0

 -- Jan Grewe <jan.grewe@g-node.org> Tue, 04 Oct 2016 15:12:00 +0100

nix (1.2.0)  trusty; urgency=medium

 * Set to version 1.2.0
 * Dimensions are no longer treated as a stack. Removing them is only possible at once.
 * Added inverse search from metadata to data entities
 * Ready for hdf5 version 1.10
 * safety checks to avoid integer overflow

 -- Jan Grewe <jan.grewe@g-node.org> Sat, 01 Oct 2016 18:18:00 +0100

nix (1.1.0-5) xenial; urgency=medium

  * Set to version 1.1.0

 -- Jan Grewe <jan.grewe@g-node.org>  Tue, 27 Jan 2016 09:45:18 +0100

nix (1.1.0-4) trusty; urgency=medium

  * Set to version 1.1.0

 -- Jan Grewe <jan.grewe@g-node.org>  Tue, 26 Jan 2016 16:31:18 -0500

nix (1.0.2-2) UNRELEASED; urgency=medium

  * debian/copyright is DEP-5 format now
  * debian/control
    - boosted policy to 3.9.6
    - fixed/extended Vcs fields
    - adjusted descriptions

 -- Yaroslav Halchenko <debian@onerussian.com>  Sat, 17 Oct 2015 10:56:29 -0500

Nix (1.0.2-1) trusty; urgency=medium

  * Set to version 1.0.2

 -- Adrian Stoewer <adrian.stoewer@rz.ifi.lmu.de>  Tue, 10 Mar 2015 09:43:50 +0100


navtree.js
var NAVTREE =
[
  [ "NIX", "index.html", [
    [ "Namespaces", null, [
      [ "Namespace List", "namespaces.html", "namespaces" ],
      [ "Namespace Members", "namespacemembers.html", [
        [ "All", "namespacemembers.html", null ],
        [ "Functions", "namespacemembers_func.html", null ],
        [ "Variables", "namespacemembers_vars.html", null ],
        [ "Typedefs", "namespacemembers_type.html", null ],
        [ "Enumerations", "namespacemembers_enum.html", null ],
        [ "Enumerator", "namespacemembers_eval.html", null ]
      ] ]
    ] ],
    [ "Classes", null, [
      [ "Class List", "annotated.html", "annotated" ],
      [ "Class Index", "classes.html", null ],
      [ "Class Hierarchy", "hierarchy.html", "hierarchy" ],
      [ "Class Members", "functions.html", [
        [ "All", "functions.html", "functions_dup" ],
        [ "Functions", "functions_func.html", "functions_func" ],
        [ "Variables", "functions_vars.html", null ],
        [ "Typedefs", "functions_type.html", null ],
        [ "Related Functions", "functions_rela.html", null ]
      ] ]
    ] ],
    [ "Files", null, [
      [ "File List", "files.html", "files" ],
      [ "File Members", "globals.html", [
        [ "All", "globals.html", null ],
        [ "Macros", "globals_defs.html", null ]
      ] ]
    ] ]
  ] ]
];

var NAVTREEINDEX =
[
"annotated.html",
"d3/d02/classnix_1_1base_1_1_i_file.html#a48011c4481118a4c77713cf9fdbdbe74",
"d5/d53/classnix_1_1_n_d_array.html#ac131f70d32465a3344b21ad06e7aaa32",
"d7/dde/classnix_1_1_property.html#ad52e348f1bcb7da4e53d38037929fd2f",
"dc/da0/classnix_1_1_source.html#abcc8183b79dd3a484297b787011f3e17",
"hierarchy.html"
];

var SYNCONMSG = 'click to disable panel synchronisation';
var SYNCOFFMSG = 'click to enable panel synchronisation';
var SYNCONMSG = 'click to disable panel synchronisation';
var SYNCOFFMSG = 'click to enable panel synchronisation';
var navTreeSubIndices = new Array();

function getData(varName)
{
  var i = varName.lastIndexOf('/');
  var n = i>=0 ? varName.substring(i+1) : varName;
  return eval(n.replace(/\-/g,'_'));
}

function stripPath(uri)
{
  return uri.substring(uri.lastIndexOf('/')+1);
}

function stripPath2(uri)
{
  var i = uri.lastIndexOf('/');
  var s = uri.substring(i+1);
  var m = uri.substring(0,i+1).match(/\/d\w\/d\w\w\/$/);
  return m ? uri.substring(i-6) : s;
}

function localStorageSupported()
{
  try {
    return 'localStorage' in window && window['localStorage'] !== null && window.localStorage.getItem;
  }
  catch(e) {
    return false;
  }
}


function storeLink(link)
{
  if (!$("#nav-sync").hasClass('sync') && localStorageSupported()) {
      window.localStorage.setItem('navpath',link);
  }
}

function deleteLink()
{
  if (localStorageSupported()) {
    window.localStorage.setItem('navpath','');
  } 
}

function cachedLink()
{
  if (localStorageSupported()) {
    return window.localStorage.getItem('navpath');
  } else {
    return '';
  }
}

function getScript(scriptName,func,show)
{
  var head = document.getElementsByTagName("head")[0]; 
  var script = document.createElement('script');
  script.id = scriptName;
  script.type = 'text/javascript';
  script.onload = func; 
  script.src = scriptName+'.js'; 
  if ($.browser.msie && $.browser.version<=8) { 
    // script.onload does not work with older versions of IE
    script.onreadystatechange = function() {
      if (script.readyState=='complete' || script.readyState=='loaded') { 
        func(); if (show) showRoot(); 
      }
    }
  }
  head.appendChild(script); 
}

function createIndent(o,domNode,node,level)
{
  var level=-1;
  var n = node;
  while (n.parentNode) { level++; n=n.parentNode; }
  if (node.childrenData) {
    var imgNode = document.createElement("img");
    imgNode.style.paddingLeft=(16*level).toString()+'px';
    imgNode.width  = 16;
    imgNode.height = 22;
    imgNode.border = 0;
    node.plus_img = imgNode;
    node.expandToggle = document.createElement("a");
    node.expandToggle.href = "javascript:void(0)";
    node.expandToggle.onclick = function() {
      if (node.expanded) {
        $(node.getChildrenUL()).slideUp("fast");
        node.plus_img.src = node.relpath+"ftv2pnode.png";
        node.expanded = false;
      } else {
        expandNode(o, node, false, false);
      }
    }
    node.expandToggle.appendChild(imgNode);
    domNode.appendChild(node.expandToggle);
    imgNode.src = node.relpath+"ftv2pnode.png";
  } else {
    var span = document.createElement("span");
    span.style.display = 'inline-block';
    span.style.width   = 16*(level+1)+'px';
    span.style.height  = '22px';
    span.innerHTML = '&#160;';
    domNode.appendChild(span);
  } 
}

var animationInProgress = false;

function gotoAnchor(anchor,aname,updateLocation)
{
  var pos, docContent = $('#doc-content');
  if (anchor.parent().attr('class')=='memItemLeft' ||
      anchor.parent().attr('class')=='fieldtype' ||
      anchor.parent().is(':header')) 
  {
    pos = anchor.parent().position().top;
  } else if (anchor.position()) {
    pos = anchor.position().top;
  }
  if (pos) {
    var dist = Math.abs(Math.min(
               pos-docContent.offset().top,
               docContent[0].scrollHeight-
               docContent.height()-docContent.scrollTop()));
    animationInProgress=true;
    docContent.animate({
      scrollTop: pos + docContent.scrollTop() - docContent.offset().top
    },Math.max(50,Math.min(500,dist)),function(){
      if (updateLocation) window.location.href=aname;
      animationInProgress=false;
    });
  }
}

function newNode(o, po, text, link, childrenData, lastNode)
{
  var node = new Object();
  node.children = Array();
  node.childrenData = childrenData;
  node.depth = po.depth + 1;
  node.relpath = po.relpath;
  node.isLast = lastNode;

  node.li = document.createElement("li");
  po.getChildrenUL().appendChild(node.li);
  node.parentNode = po;

  node.itemDiv = document.createElement("div");
  node.itemDiv.className = "item";

  node.labelSpan = document.createElement("span");
  node.labelSpan.className = "label";

  createIndent(o,node.itemDiv,node,0);
  node.itemDiv.appendChild(node.labelSpan);
  node.li.appendChild(node.itemDiv);

  var a = document.createElement("a");
  node.labelSpan.appendChild(a);
  node.label = document.createTextNode(text);
  node.expanded = false;
  a.appendChild(node.label);
  if (link) {
    var url;
    if (link.substring(0,1)=='^') {
      url = link.substring(1);
      link = url;
    } else {
      url = node.relpath+link;
    }
    a.className = stripPath(link.replace('#',':'));
    if (link.indexOf('#')!=-1) {
      var aname = '#'+link.split('#')[1];
      var srcPage = stripPath($(location).attr('pathname'));
      var targetPage = stripPath(link.split('#')[0]);
      a.href = srcPage!=targetPage ? url : "javascript:void(0)"; 
      a.onclick = function(){
        storeLink(link);
        if (!$(a).parent().parent().hasClass('selected'))
        {
          $('.item').removeClass('selected');
          $('.item').removeAttr('id');
          $(a).parent().parent().addClass('selected');
          $(a).parent().parent().attr('id','selected');
        }
        var anchor = $(aname);
        gotoAnchor(anchor,aname,true);
      };
    } else {
      a.href = url;
      a.onclick = function() { storeLink(link); }
    }
  } else {
    if (childrenData != null) 
    {
      a.className = "nolink";
      a.href = "javascript:void(0)";
      a.onclick = node.expandToggle.onclick;
    }
  }

  node.childrenUL = null;
  node.getChildrenUL = function() {
    if (!node.childrenUL) {
      node.childrenUL = document.createElement("ul");
      node.childrenUL.className = "children_ul";
      node.childrenUL.style.display = "none";
      node.li.appendChild(node.childrenUL);
    }
    return node.childrenUL;
  };

  return node;
}

function showRoot()
{
  var headerHeight = $("#top").height();
  var footerHeight = $("#nav-path").height();
  var windowHeight = $(window).height() - headerHeight - footerHeight;
  (function (){ // retry until we can scroll to the selected item
    try {
      var navtree=$('#nav-tree');
      navtree.scrollTo('#selected',0,{offset:-windowHeight/2});
    } catch (err) {
      setTimeout(arguments.callee, 0);
    }
  })();
}

function expandNode(o, node, imm, showRoot)
{
  if (node.childrenData && !node.expanded) {
    if (typeof(node.childrenData)==='string') {
      var varName    = node.childrenData;
      getScript(node.relpath+varName,function(){
        node.childrenData = getData(varName);
        expandNode(o, node, imm, showRoot);
      }, showRoot);
    } else {
      if (!node.childrenVisited) {
        getNode(o, node);
      } if (imm || ($.browser.msie && $.browser.version>8)) { 
        // somehow slideDown jumps to the start of tree for IE9 :-(
        $(node.getChildrenUL()).show();
      } else {
        $(node.getChildrenUL()).slideDown("fast");
      }
      if (node.isLast) {
        node.plus_img.src = node.relpath+"ftv2mlastnode.png";
      } else {
        node.plus_img.src = node.relpath+"ftv2mnode.png";
      }
      node.expanded = true;
    }
  }
}

function glowEffect(n,duration)
{
  n.addClass('glow').delay(duration).queue(function(next){
    $(this).removeClass('glow');next();
  });
}

function highlightAnchor()
{
  var aname = $(location).attr('hash');
  var anchor = $(aname);
  if (anchor.parent().attr('class')=='memItemLeft'){
    var rows = $('.memberdecls tr[class$="'+
               window.location.hash.substring(1)+'"]');
    glowEffect(rows.children(),300); // member without details
  } else if (anchor.parents().slice(2).prop('tagName')=='TR') {
    glowEffect(anchor.parents('div.memitem'),1000); // enum value
  } else if (anchor.parent().attr('class')=='fieldtype'){
    glowEffect(anchor.parent().parent(),1000); // struct field
  } else if (anchor.parent().is(":header")) {
    glowEffect(anchor.parent(),1000); // section header
  } else {
    glowEffect(anchor.next(),1000); // normal member
  }
  gotoAnchor(anchor,aname,false);
}

function selectAndHighlight(hash,n)
{
  var a;
  if (hash) {
    var link=stripPath($(location).attr('pathname'))+':'+hash.substring(1);
    a=$('.item a[class$="'+link+'"]');
  }
  if (a && a.length) {
    a.parent().parent().addClass('selected');
    a.parent().parent().attr('id','selected');
    highlightAnchor();
  } else if (n) {
    $(n.itemDiv).addClass('selected');
    $(n.itemDiv).attr('id','selected');
  }
  if ($('#nav-tree-contents .item:first').hasClass('selected')) {
    $('#nav-sync').css('top','30px');
  } else {
    $('#nav-sync').css('top','5px');
  }
  showRoot();
}

function showNode(o, node, index, hash)
{
  if (node && node.childrenData) {
    if (typeof(node.childrenData)==='string') {
      var varName    = node.childrenData;
      getScript(node.relpath+varName,function(){
        node.childrenData = getData(varName);
        showNode(o,node,index,hash);
      },true);
    } else {
      if (!node.childrenVisited) {
        getNode(o, node);
      }
      $(node.getChildrenUL()).css({'display':'block'});
      if (node.isLast) {
        node.plus_img.src = node.relpath+"ftv2mlastnode.png";
      } else {
        node.plus_img.src = node.relpath+"ftv2mnode.png";
      }
      node.expanded = true;
      var n = node.children[o.breadcrumbs[index]];
      if (index+1<o.breadcrumbs.length) {
        showNode(o,n,index+1,hash);
      } else {
        if (typeof(n.childrenData)==='string') {
          var varName = n.childrenData;
          getScript(n.relpath+varName,function(){
            n.childrenData = getData(varName);
            node.expanded=false;
            showNode(o,node,index,hash); // retry with child node expanded
          },true);
        } else {
          var rootBase = stripPath(o.toroot.replace(/\..+$/, ''));
          if (rootBase=="index" || rootBase=="pages" || rootBase=="search") {
            expandNode(o, n, true, true);
          }
          selectAndHighlight(hash,n);
        }
      }
    }
  } else {
    selectAndHighlight(hash);
  }
}

function removeToInsertLater(element) {
  var parentNode = element.parentNode;
  var nextSibling = element.nextSibling;
  parentNode.removeChild(element);
  return function() {
    if (nextSibling) {
      parentNode.insertBefore(element, nextSibling);
    } else {
      parentNode.appendChild(element);
    }
  };
}

function getNode(o, po)
{
  var insertFunction = removeToInsertLater(po.li);
  po.childrenVisited = true;
  var l = po.childrenData.length-1;
  for (var i in po.childrenData) {
    var nodeData = po.childrenData[i];
    po.children[i] = newNode(o, po, nodeData[0], nodeData[1], nodeData[2],
      i==l);
  }
  insertFunction();
}

function gotoNode(o,subIndex,root,hash,relpath)
{
  var nti = navTreeSubIndices[subIndex][root+hash];
  o.breadcrumbs = $.extend(true, [], nti ? nti : navTreeSubIndices[subIndex][root]);
  if (!o.breadcrumbs && root!=NAVTREE[0][1]) { // fallback: show index
    navTo(o,NAVTREE[0][1],"",relpath);
    $('.item').removeClass('selected');
    $('.item').removeAttr('id');
  }
  if (o.breadcrumbs) {
    o.breadcrumbs.unshift(0); // add 0 for root node
    showNode(o, o.node, 0, hash);
  }
}

function navTo(o,root,hash,relpath)
{
  var link = cachedLink();
  if (link) {
    var parts = link.split('#');
    root = parts[0];
    if (parts.length>1) hash = '#'+parts[1];
    else hash='';
  }
  if (hash.match(/^#l\d+$/)) {
    var anchor=$('a[name='+hash.substring(1)+']');
    glowEffect(anchor.parent(),1000); // line number
    hash=''; // strip line number anchors
    //root=root.replace(/_source\./,'.'); // source link to doc link
  }
  var url=root+hash;
  var i=-1;
  while (NAVTREEINDEX[i+1]<=url) i++;
  if (i==-1) { i=0; root=NAVTREE[0][1]; } // fallback: show index
  if (navTreeSubIndices[i]) {
    gotoNode(o,i,root,hash,relpath)
  } else {
    getScript(relpath+'navtreeindex'+i,function(){
      navTreeSubIndices[i] = eval('NAVTREEINDEX'+i);
      if (navTreeSubIndices[i]) {
        gotoNode(o,i,root,hash,relpath);
      }
    },true);
  }
}

function showSyncOff(n,relpath)
{
    n.html('<img src="'+relpath+'sync_off.png" title="'+SYNCOFFMSG+'"/>');
}

function showSyncOn(n,relpath)
{
    n.html('<img src="'+relpath+'sync_on.png" title="'+SYNCONMSG+'"/>');
}

function toggleSyncButton(relpath)
{
  var navSync = $('#nav-sync');
  if (navSync.hasClass('sync')) {
    navSync.removeClass('sync');
    showSyncOff(navSync,relpath);
    storeLink(stripPath2($(location).attr('pathname'))+$(location).attr('hash'));
  } else {
    navSync.addClass('sync');
    showSyncOn(navSync,relpath);
    deleteLink();
  }
}

function initNavTree(toroot,relpath)
{
  var o = new Object();
  o.toroot = toroot;
  o.node = new Object();
  o.node.li = document.getElementById("nav-tree-contents");
  o.node.childrenData = NAVTREE;
  o.node.children = new Array();
  o.node.childrenUL = document.createElement("ul");
  o.node.getChildrenUL = function() { return o.node.childrenUL; };
  o.node.li.appendChild(o.node.childrenUL);
  o.node.depth = 0;
  o.node.relpath = relpath;
  o.node.expanded = false;
  o.node.isLast = true;
  o.node.plus_img = document.createElement("img");
  o.node.plus_img.src = relpath+"ftv2pnode.png";
  o.node.plus_img.width = 16;
  o.node.plus_img.height = 22;

  if (localStorageSupported()) {
    var navSync = $('#nav-sync');
    if (cachedLink()) {
      showSyncOff(navSync,relpath);
      navSync.removeClass('sync');
    } else {
      showSyncOn(navSync,relpath);
    }
    navSync.click(function(){ toggleSyncButton(relpath); });
  }

  $(window).load(function(){
    navTo(o,toroot,window.location.hash,relpath);
    showRoot();
  });

  $(window).bind('hashchange', function(){
     if (window.location.hash && window.location.hash.length>1){
       var a;
       if ($(location).attr('hash')){
         var clslink=stripPath($(location).attr('pathname'))+':'+
                               $(location).attr('hash').substring(1);
         a=$('.item a[class$="'+clslink+'"]');
       }
       if (a==null || !$(a).parent().parent().hasClass('selected')){
         $('.item').removeClass('selected');
         $('.item').removeAttr('id');
       }
       var link=stripPath2($(location).attr('pathname'));
       navTo(o,link,$(location).attr('hash'),relpath);
     } else if (!animationInProgress) {
       $('#doc-content').scrollTop(0);
       $('.item').removeClass('selected');
       $('.item').removeAttr('id');
       navTo(o,toroot,window.location.hash,relpath);
     }
  })
}

